// util/output_listener.js
// Copyright Titanium I.T. LLC.
import * as ensure from "util/ensure.js";
import * as types from "util/type.js";
import EventEmitter from "node:events";

const EVENT = "event";

export class OutputListener {

	static create() {
		ensure.signature(arguments, []);

		return new OutputListener();
	}

	constructor() {
		this._emitter = new EventEmitter();
	}

	trackOutput() {
		ensure.signature(arguments, []);

		return new OutputTracker(this._emitter, EVENT);
	}

	emit(data) {
		ensure.signature(arguments, [ types.ANY ]);

		this._emitter.emit(EVENT, data);
	}

}


class OutputTracker {

	constructor(emitter, event) {
		this._emitter = emitter;
		this._event = event;
		this._data = [];

		this._trackerFn = (text) => this._data.push(text);
		this._emitter.on(this._event, this._trackerFn);
	}

	get data() {
		return this._data;
	}

	clear() {
		ensure.signature(arguments, []);

		const result = [ ...this._data ];
		this._data.length = 0;
		return result;
	}

	stop() {
		ensure.signature(arguments, []);

		this._emitter.off(this._event, this._trackerFn);
	}

}
